pkg/tls/tls.go

```diff
@@ -0,0 +1,149 @@
+	}
+
+	mngr, err := newTLSManager(srvCertPath, srvKeyPath, clientCApath)
+	if err != nil {
+		return nil, err
+	}
+	tlsCfg.GetCertificate = mngr.getCertificate
+	level.Info(logger).Log("msg", "enabled gRPC server side TLS")
+
+	if clientCApath != "" {
+		tlsCfg.ClientAuth = tls.RequireAndVerifyClientCert
+		tlsCfg.GetClientCertificate = mngr.getClientCertificate
+		tlsCfg.ClientCAs = mngr.certPool // Send a hint to the client which are the trusted CAs.
+		level.Info(logger).Log("msg", "gRPC server TLS client verification enabled")
+	}
+
+	return tlsCfg, nil
+}
+
+type tlsManager struct {
+	srvCertPath,
+	srvKeyPath,
+	cltCApath string
+
+	ticker <-chan time.Time
+
+	mtxSrv         sync.Mutex
+	srvCert        *tls.Certificate
+	srvCertModTime time.Time
+	srvKeyModTime  time.Time
+
+	mtxClt       sync.Mutex
+	cltCA        *tls.Certificate
+	certPool     *x509.CertPool
+	cltCAmodTime time.Time
+}
+
+func newTLSManager(srvCertPath, srvKeyPath, cltCApath string) (*tlsManager, error) {
+	m := &tlsManager{
+		srvCertPath: srvCertPath,
+		srvKeyPath:  srvKeyPath,
+		cltCApath:   cltCApath,
+		ticker:      time.NewTicker(1000 * time.Millisecond).C,
+	}
+
+	return m, nil
+}
+
+func (m *tlsManager) getCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	// Throttle updates to once per second.
+	select {
+	case <-m.ticker:
+	default:
+		if m.srvCert == nil {
+			return m.srvCert, nil
+		}
+	}
+
+	statCert, err := os.Stat(m.srvCertPath)
+	if err != nil {
+		return nil, err
+	}
+	statKey, err := os.Stat(m.srvKeyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	m.mtxSrv.Lock()
+	defer m.mtxSrv.Unlock()
+
+	if m.srvCert == nil || !statCert.ModTime().Equal(m.srvCertModTime) || !statKey.ModTime().Equal(m.srvKeyModTime) {
+		cert, err := tls.LoadX509KeyPair(m.srvCertPath, m.srvKeyPath)
+		if err != nil {
+			return nil, errors.Wrap(err, "loading server cert and key")
+		}
+		m.srvCertModTime = statCert.ModTime()
+		m.srvKeyModTime = statKey.ModTime()
+		m.srvCert = &cert
+	}
+	return m.srvCert, nil
+}
+func (m *tlsManager) getClientCertificate(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+	// Throttle updates to once per second.
+	select {
+	case <-m.ticker:
+	default:
+		if m.cltCA != nil {
+			return m.cltCA, nil
+		}
+	}
+
+	statCA, err := os.Stat(m.cltCApath)
+	if err != nil {
+		return nil, err
+	}
+
+	m.mtxClt.Lock()
+	defer m.mtxClt.Unlock()
+
+	if m.cltCA == nil || !statCA.ModTime().Equal(m.cltCAmodTime) {
+		caPEM, err := ioutil.ReadFile(m.cltCApath)
+		if err != nil {
+			return nil, errors.Wrap(err, "reading client CA")
+		}
+
+		certPool := x509.NewCertPool()
+		if !certPool.AppendCertsFromPEM(caPEM) {
+			return nil, errors.New("adding CA file to the client pool")
+		}
+
+		caCert, err := tls.X509KeyPair(caPEM, caPEM)
+		if err != nil {
+			return nil, errors.Wrap(err, "building client CA certificate")
+		}
+
+		// ????? test if true...... This is a pointer it will update the underlying tls config pool.
```

Comment:
I guess this is still a work in progress because of these comments?

Reply:
yep one of the items in my todo list
> test if the CA pool gets updated with the ponter